daft/table/table_io.py

```diff
@@ -663,6 +665,189 @@ def file_visitor(written_file, protocol=protocol):
     return MicroPartition.from_pydict({"data_file": Series.from_pylist(data_files, name="data_file", pyobj="force")})
 
 
+def write_deltalake(
+    mp: MicroPartition,
+    large_dtypes: bool,
+    base_path: str,
+    current_version: int,
+    mode: str,
+    invariants: list[tuple[str, str]] | None,
+    delta_table: DeltaTable | None,
+    file_writer_spec: list[tuple[str, int | None]],
+    partition_filters: list[tuple[str, str, Any]] | None,
+    partition_by: list[str] | None,
+    io_config: IOConfig | None = None,
+):
+
+    from deltalake.schema import convert_pyarrow_table
+    from deltalake.writer import (
+        AddAction,
+        DeltaJSONEncoder,
+        DeltaStorageHandler,
+        _DeltaDataChecker,
+        batch_distinct,
+        encode_partition_value,
+        get_file_stats_from_metadata,
+        get_partitions_from_path,
+        try_get_table_and_table_uri,
+    )
+    from pyarrow import dataset as ds
+    from pyarrow.fs import PyFileSystem
+    from pyarrow.lib import RecordBatchReader
+
+    from daft.delta_lake.delta_lake_storage_function import (
+        _storage_config_to_storage_options,
+    )
+
+    data_files: list[AddAction] = []
+
+    def file_visitor(written_file: Any) -> None:
+        path, partition_values = get_partitions_from_path(written_file.path)
+        stats = get_file_stats_from_metadata(written_file.metadata)
+
+        import json
+        from datetime import datetime
+
+        from daft.utils import ARROW_VERSION
+
+        # PyArrow added support for written_file.size in 9.0.0
+        if ARROW_VERSION >= (9, 0, 0):
+            size = written_file.size
+        elif filesystem is not None:
+            size = filesystem.get_file_info([path])[0].size
+        else:
+            size = 0
+
+        data_files.append(
+            AddAction(
+                path,
+                size,
+                partition_values,
+                int(datetime.now().timestamp() * 1000),
+                True,
+                json.dumps(stats, cls=DeltaJSONEncoder),
+            )
+        )
+
+    io_config = get_context().daft_planning_config.default_io_config if io_config is None else io_config
+    storage_config = StorageConfig.native(NativeStorageConfig(False, io_config))
+    storage_options = _storage_config_to_storage_options(storage_config, base_path)
+    table, table_uri = try_get_table_and_table_uri(base_path, storage_options)
+    filesystem = PyFileSystem(DeltaStorageHandler(table_uri, storage_options))
+
+    arrow_table = mp.to_arrow()
+    arrow_batch = convert_pyarrow_table(arrow_table, large_dtypes)
+    schema = arrow_batch.schema
+
+    import deltalake
+    from packaging.version import parse
+
+    Deltaschema = None
+    if parse(deltalake.__version__) < parse("0.14.0"):
```

Comment:
changed the requirement*.txt . 
but getting weird error in github action 🤧
```INTERNALERROR> AssertionError: Expected current collector to be <Collector at 0x7f9e63981e90: CTracer>, but it's <Collector at 0x7f9e5011f590: CTracer>```

Reply:
Ah I see! That's usually an import error. I merged in this PR to raise the import error in CI. You should be able to rebase and see what is happening.
https://github.com/Eventual-Inc/Daft/pull/2150

Reply:
weird !. 
looks like pytest is not finding in any test during 3.7 ubunutu  build .
i have removed this line 
```deltalake==0.13.0; platform_system != "Windows" and python_version < '3.8'```
 from requirement-dev.txt because we can't support write with 0.13.0 without the above code .
 not able to build currently with python 3.7 in local also .